Similarity/sim_functions.R

```diff
@@ -86,205 +84,62 @@ dynlib_map <- function(x, y, xPrecursorMz, yPrecursorMz, ...) {
   mz1 <- cleaned1$mz
   intensity1 <- cleaned1$intensity
   mz2 <- cleaned2$mz
   intensity2 <- cleaned2$intensity
-<<<<<<< HEAD
-  
-  # Matching exact m/z
-  common_mz <- intersect(mz1, mz2)
-  matched1 <- cbind(mz1[mz1 %in% common_mz], intensity1[mz1 %in% common_mz])
-  matched2 <- cbind(mz2[mz2 %in% common_mz], intensity2[mz2 %in% common_mz])
-  
-  # Remaining peaks
-  remaining_idx1 <- !(mz1 %in% common_mz)
-  remaining_idx2 <- !(mz2 %in% common_mz)
-  
-  nl1 <- neutral_loss(mz1[remaining_idx1], xPrecursorMz)
-  nl2 <- neutral_loss(mz2[remaining_idx2], yPrecursorMz)
-  common_nl <- intersect(nl1, nl2)
-  
-  for (nl in common_nl) {
-    idx_nl1 <- which(nl1 == nl)
-    idx_nl2 <- which(nl2 == nl)
-    
-    matched_nl1 <- cbind(mz1[remaining_idx1][idx_nl1],
-                         intensity1[remaining_idx1][idx_nl1])
-    matched_nl2 <- cbind(mz2[remaining_idx2][idx_nl2],
-                         intensity2[remaining_idx2][idx_nl2])
-    
-    matched1 <- rbind(matched1, matched_nl1)
-    matched2 <- rbind(matched2, matched_nl2)
-  }
-  
-  return(list(x = matched1, y = matched2))
-=======
-
-  intensity1 <- (intensity1^n) * (mz1^m)
-  intensity2 <- (intensity2^n) * (mz2^m)
-
-  Fd_denom1 <- sum(intensity1^2)
-  Fd_denom2 <- sum(intensity2^2)
 
+  # Matching exact m/z
   common_mz <- intersect(mz1, mz2)
-  Fd_nom <- 0
-  remaining_idx1 <- rep(TRUE, length(mz1))
-  remaining_idx2 <- rep(TRUE, length(mz2))
-
-  if (length(common_mz) > 0) {
-    idx1 <- match(common_mz, mz1)
-    idx2 <- match(common_mz, mz2)
-
-    valid_indices <- !is.na(idx1) & !is.na(idx2)
-    idx1 <- idx1[valid_indices]
-    idx2 <- idx2[valid_indices]
-
-    if (length(idx1) > 0 && length(idx2) > 0) {
-      Fd_nom <- sum(intensity1[idx1] * intensity2[idx2])
-    }
-
-    remaining_idx1[idx1] <- FALSE
-    remaining_idx2[idx2] <- FALSE
-  }
-
-  # Calcul des neutral losses sur les pics restants
-  nl1 <- neutral_loss(mz1[remaining_idx1], xPrecursorMz)
-  nl2 <- neutral_loss(mz2[remaining_idx2], yPrecursorMz)
-
-  common_nl <- intersect(nl1, nl2)
-
-  if (length(common_nl) > 0) {
-    for (nl in common_nl) {
-      matches1 <- which(nl1 == nl)
-      matches2 <- which(nl2 == nl)
-
-      if (length(matches1) > 0 && length(matches2) > 0) {
-        Fd_nom <- Fd_nom + sum(intensity1[remaining_idx1][matches1]) *
-          sum(intensity2[remaining_idx2][matches2])
+  keep_mz1 <- mz1 %in% common_mz
+  keep_mz2 <- mz2 %in% common_mz
+  matched1 <- cbind(mz = mz1[keep_mz1], intensity = intensity1[keep_mz1])
+  matched2 <- cbind(mz = mz2[keep_mz2], intensity = intensity2[keep_mz2])
+
+  ## Process the remaining peaks to see if they could match as neutral losses
+  remaining_mz1 <- !keep_mz1
+  remaining_mz2 <- !keep_mz2
+  if (any(remaining_mz1) && any(remaining_mz2)) {
+      nl1 <- neutral_loss(mz1[remaining_mz1], xPrecursorMz)
+      nl2 <- neutral_loss(mz2[remaining_mz2], yPrecursorMz)
+      common_nl <- intersect(nl1, nl2)
+      if (length(common_nl)) {
```

Comment:
also here, only do this if we have common neutral loss fragment peaks. and we can use the same approach as above, no need for a `for` loop